src/mapping/query/function/split.rs

```diff
@@ -0,0 +1,183 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct SplitFn {
+    path: Box<dyn Function>,
+    pattern: ArgumentKind,
+    limit: Option<Box<dyn Function>>,
+}
+
+impl SplitFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        path: Box<dyn Function>,
+        pattern: ArgumentKind,
+        limit: Option<Box<dyn Function>>,
+    ) -> Self {
+        Self {
+            path,
+            pattern,
+            limit,
+        }
+    }
+}
+
+impl Function for SplitFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let string = {
+            let bytes = required!(ctx, self.path, Value::Bytes(v) => v);
+            String::from_utf8_lossy(&bytes).into_owned()
+        };
+
+        let limit = optional!(ctx, self.limit, Value::Integer(i) => i)
+            .map(|limit| {
+                if limit < 0 {
+                    Err("limit is not a positive number".to_string())
+                } else {
+                    Ok(limit as usize)
+                }
+            })
+            .transpose()?;
+
+        match &self.pattern {
+            ArgumentKind::Value(path) => {
+                let pattern = required!(ctx, path, Value::Bytes(v) => v);
+                let pattern = String::from_utf8_lossy(&pattern).into_owned();
+
+                let iter: Box<dyn Iterator<Item = _>> = match limit {
+                    Some(limit) => Box::new(string.splitn(limit, &pattern)),
+                    None => Box::new(string.split(&pattern)),
+                };
+
+                Ok(Value::Array(
+                    iter.map(|sub| Value::Bytes(sub.to_string().into()))
+                        .collect(),
+                ))
```

Comment:
Seems like we should be able to pull this (and below) out of the match statement to reduce duplication.

Reply:
Seconded. I think all of the iterators should even be the same type thanks to the box, so everything except the split could move up.

Reply:
Unfortunately, this gives me lifetime issues. The pattern and regex are both dropped as we leave the `match`, but the iterator still needs them to be around as it iterates through the results.

The best I can do is extract the repeated code out to a closure and call that from each branch.